packages/flutter/lib/src/services/platform_messages.dart

```diff
@@ -237,3 +250,647 @@ class PlatformMessages {
+/// populate a [PlatformException].
+///
+/// All operations throw [FormatException], if conversion fails.
+abstract class MessageCodec {
+  /// [MessageCodec] with UTF-8 encoded JSON messages.
+  ///
+  /// Supported messages are acyclic values of these forms:
+  ///
+  /// * `null`
+  /// * [bool]s
+  /// * [num]s
+  /// * [String]s
+  /// * [List]s of supported values
+  /// * [Map]s from strings to supported values
+  static const MessageCodec json = const _JSONCodec();
+
+  /// [MessageCodec] using the Flutter standard binary message encoding.
+  ///
+  /// The standard codec is guaranteed to be compatible with the corresponding
+  /// standard codec for FlutterChannels on the host platform. These two parts
+  /// of the Flutter SDK are evolved synchronously.
+  ///
+  /// Supported messages are acyclic values of these forms:
+  ///
+  /// * `null`
+  /// * [bool]s
+  /// * [num]s
+  /// * [String]s
+  /// * [Uint8List]s, [Int32List]s, [Int64List]s, [Float64List]s
+  /// * [List]s of supported values
+  /// * [Map]s from supported values to supported values
+  static const MessageCodec standard = const _StandardCodec();
+
+  /// Encodes the specified [message] in binary.
+  ///
+  /// Returns `null` if the message is `null`.
+  ByteData encodeMessage(dynamic message);
+
+  /// Decodes the specified [message] from binary.
+  ///
+  /// Returns `null` if the message is `null`.
+  dynamic decodeMessage(ByteData message);
+
+  /// Encodes the specified method call in binary.
+  ///
+  /// The [name] of the method must be non-null. The [arguments] may be `null`.
+  ByteData encodeMethodCall(String name, dynamic arguments);
+
+  /// Decodes the specified reply [envelope] from binary.
+  ///
+  /// Throws [PlatformException], if [envelope] represents an error.
+  dynamic decodeEnvelope(ByteData envelope);
+}
+
+class _JSONCodec implements MessageCodec {
+  const _JSONCodec();
+
+  @override
+  ByteData encodeMessage(dynamic message) => _encodeUTF8(_encodeJSON(message));
+
+  @override
+  dynamic decodeMessage(ByteData message) => _decodeJSON(_decodeUTF8(message));
+
+  @override
+  ByteData encodeMethodCall(String name, dynamic arguments) {
+    assert(name != null);
+    return encodeMessage(<String, dynamic>{
+      'method': name,
+      'args': arguments,
+    });
+  }
+
+  @override
+  dynamic decodeEnvelope(ByteData envelope) {
+    final dynamic decoded = decodeMessage(envelope);
+    if (decoded is! Map)
+      throw new FormatException('Expected envelope Map, got $decoded');
+    final String status = decoded['status'];
+    final dynamic data = decoded['data'];
+    if (status == 'ok')
+      return data;
+    final String message = decoded['message'];
+    if (status is String && message is String)
+      throw new PlatformException(code: status, message: message, details: data);
+    throw new FormatException('Invalid envelope $decoded');
+  }
+}
+
+class _StandardCodec implements MessageCodec {
```

Comment:
As requested in the Java review, please document this protocol.  Here are examples where we've documented similar protocols:

https://github.com/flutter/engine/blob/master/lib/ui/text.dart#L184
https://github.com/flutter/engine/blob/master/lib/ui/painting.dart#L310

Doesn't have to be anything fancy.  Just enough documentation so that someone could write, for example, a C# implementation of the protocol without having to reverse engineer the other implementations.